packages/go_router/lib/src/delegate.dart

```diff
@@ -150,20 +150,43 @@ class GoRouterDelegate extends RouterDelegate<RouteMatchList>
 
   bool _handlePopPageWithRouteMatch(
       Route<Object?> route, Object? result, RouteMatch? match) {
-    if (!route.didPop(result)) {
-      return false;
+    if (route.willHandlePopInternally) {
+      final bool popped = route.didPop(result);
+      assert(!popped);
+      return popped;
     }
     assert(match != null);
+    if (match!.route.onExit != null) {
+      // The `onExit` may perform updates to the navigator, such as popping up a
+      // dialog. Use a microtask to ensure the navigator finishes the current
+      // update before calling the `onExit`.
+      scheduleMicrotask(() async {
+        final bool onExitResult =
+            await match.route.onExit!(navigatorKey.currentContext!);
+        if (onExitResult) {
+          _removeMatchFromList(match, result);
+        }
+      });
+      return false;
+    } else {
+      if (!route.didPop(result)) {
+        return false;
+      }
+      _removeMatchFromList(match, result);
+      return true;
+    }
+  }
+
+  void _removeMatchFromList(RouteMatch match, Object? result) {
     if (match is ImperativeRouteMatch) {
       match.complete(result);
     }
-    _remove(match!);
+    _remove(match);
```

Comment:
This should also needs to handle match is no longer in match list. What would happen if while onExit is still running but another context.go or pop is called. Which onExit should take priority and what if one finished before the other? or should this throws error instead?

Reply:
Should we just `return` [here](https://github.com/flutter/packages/blob/3c3c980df1add6b34c49e8c5d712484287fa8976/packages/go_router/lib/src/matching.dart#L157) if the match cannot be found?
It means we would have already been through this function.

Reply:
It depends on how are we going to handle conflicts. The easiest way is to block navigation until onExit is finished. This way we can guarantee the page should still be here.

If we are to accept multiple calls to onExit, we may need to figure out a way to handle cases like one call return true but another returns false.

Reply:
So how do we make the decision?
If we go with the easiest approach, I can create lock in GoRouter and use it in `go`, `push`, `pushReplacement`, `replace` and `pop`. Is that what you have in mind?

Reply:
I vote for that one, since it is the easiest. We can always allow that in the future if we figure out how to handle conflicts.

One thing we need to double check is that the popRoute can still be called because we can't stop people from pressing android back button. One idea is to just return false in the onPopPage callback if the previous onExit has not completed.

we also want to think about how to handle browser forward and backward button. This one I am not entirely sure what we should do here. You can't really stop that.